coap/adapter.go

```diff
@@ -1,61 +1,257 @@
-package adapter
+package coap
 
 import (
 	"encoding/json"
-	"log"
+	"errors"
+	"fmt"
 	"net"
+	"strings"
 
-	"github.com/dustin/go-coap"
+	"encoding/base64"
+
+	mux "github.com/dereulenspiegel/coap-mux"
+	coap "github.com/dustin/go-coap"
+	"github.com/go-kit/kit/log"
 	"github.com/mainflux/mainflux"
+	manager "github.com/mainflux/mainflux/manager/client"
 	broker "github.com/nats-io/go-nats"
-	"go.uber.org/zap"
 )
 
-const protocol string = "coap"
+const (
+	token    string = "token"
+	prefix   string = "src.coap."
+	channel  string = "channel_id"
+	protocol string = "coap"
+)
+
+var (
+	errUnauthorizedAccess error = errors.New("missing or invalid credentials provided")
+)
 
+// Observer struct provides support to observe message types.
 type Observer struct {
 	conn    *net.UDPConn
 	addr    *net.UDPAddr
 	message *coap.Message
+	sub     *broker.Subscription
 }
 
-type CoAPAdapter struct {
+// Subscriber represents token-NATS subscription map. Token is represented as string in
+// order to simplify access to subscription as well as to simplify unsubscribe feature.
+type Subscriber map[string]*Observer
+
+// AdapterService struct represents CoAP adapter service.
+type AdapterService struct {
 	obsMap map[string][]Observer
-	logger *zap.Logger
+	logger log.Logger
 	pub    mainflux.MessagePublisher
+	subs   map[string]Subscriber
+	nc     *broker.Conn
+	mc     *manager.ManagerClient
 }
 
-// NewCoAPAdapter creates new CoAP adapter struct
-func NewCoAPAdapter(logger *zap.Logger, pub mainflux.MessagePublisher) *CoAPAdapter {
-	ca := &CoAPAdapter{
+// New creates new CoAP adapter service struct.
+func New(logger log.Logger, pub mainflux.MessagePublisher, nc *broker.Conn, mc *manager.ManagerClient) *AdapterService {
+	ca := &AdapterService{
 		logger: logger,
 		pub:    pub,
 		obsMap: make(map[string][]Observer),
+		subs:   make(map[string]Subscriber),
+		nc:     nc,
+		mc:     mc,
 	}
-
 	return ca
 }
 
-// Serve function starts CoAP server
-func (ca *CoAPAdapter) Serve(addr string) error {
-	ca.logger.Info("Starting CoAP server", zap.String("address", addr))
-	return coap.ListenAndServe("udp", addr, ca.COAPServer())
+// SendMessage method sends processes message and pushes it to NATS.
+func (ca *AdapterService) SendMessage(l *net.UDPConn, a *net.UDPAddr, m *coap.Message) *coap.Message {
+	ca.logger.Log("info", fmt.Sprintf("Got message in sendMessage: path=%q: %#v from %v", m.Path(), m, a))
+	var res *coap.Message
+	if len(m.Payload) == 0 && m.IsConfirmable() {
+		res.Code = coap.BadRequest
+		return res
+	}
+
+	// Uri-Query option ID is 15 (0xf).
+	token, err := ca.token(m.Option(0xf))
+	if err != nil {
+		res.Code = coap.Unauthorized
+		return res
+	}
+
+	cid := prefix + mux.Var(m, channel)
+
+	publisher, err := ca.mc.CanAccess(cid, token)
```

Comment:
Looks like CoAP token is treated like JWT

Reply:
Agree, I will do some refactoring to make it clear which token is used here.

Reply:
Since we use JWT-based auth, I found passing JWT as url query-param is pretty good workaround. Since token is wrapped as message option, it's not completely exposed. Also, obtaining token from message is straightforward and message payload stays clean. Possible problem is CoAP limitation of url query param length.

Reply:
Yes, `Uri-Query` option field looks like the good candidate for JWT passing. Let's start with this and we'll see later.

Also, maybe we should note in the comment the max size of JWT allowed (alongside with some clarifications - that we use `Uri-Query`, etc).

Reply:
Agree, I will add some clarifications as I refactor code.